Uabc 178 creation of edit schedules dashboard

tulza opened the pull request:
### Closes #178 

```created new route``` /admin/edit-schedule
```following``` 'Edit  Sessions'deign on figma
```has``` new route, back button,  session cards, popover selection.

> working in progress:
> getting real data to display,  edit and delete session dialogue.

tulza committed:
crazy radio use

tulza committed:
init Popover

tulza committed:
rename route, link admin dashboard to edit session

tulza committed:
init dialogue

tulza committed:
shadcn dialog implementation

tulza committed:
delete dialog

tulza committed:
refactor editbutton.tsx to use editdialog

tulza committed:
temp refactor of textinput to work with hook form

tulza committed:
hook form testing

dyzhuu commented on src/components/admin/EditButton.tsx:
```diff
@@ -0,0 +1,23 @@
+          className="hover:bg-secondary h-8 text-foreground justify-start w-full"
+        >
+          <SquarePen className="w-4 mr-2" />
+          <p>Edit</p>s
```
```suggestion
          <p>Edit</p>
```

tulza committed:
further textinput refactor: error message now works

dyzhuu commented on src/app/admin/edit-sessions/page.tsx:
```diff
@@ -0,0 +1,41 @@
+    <div className="relative flex px-4 min-h-dvh max-w-dvw flex-col bg-background dark overflow-x-hidden">
+      <EditSessionsHeader />
+      <div className="gap-4 flex flex-col mb-4">
+        <SessionCard />
```
These cards should actually be links

dyzhuu commented on src/components/admin/SessionCard.tsx:
```diff
@@ -0,0 +1,62 @@
+import DeleteButton from "./DeleteButton";
+import { EditButton } from "./EditButton";
+
+export const SessionCard = () => {
```
Perhaps a better name is `SemesterDetailCard`?

dyzhuu commented on src/components/admin/SessionCard.tsx:
```diff
@@ -0,0 +1,62 @@
+          Semester 2 (2024)
+        </h3>
+
+        <Button variant="outline" className="w-8 h-6 z-10" onClick={handleOpen}>
+          <Ellipsis className="stroke-tertiary absolute w-4" />
+        </Button>
+        {open && <Popover />}
```
This button and popover component will actually be reused in a lot of places. Think about how we can make this more reusable, and also be able to adjust the edit button form content & logic, as well as for delete.

tulza committed:
minor style changes + refactoring popover to be reusable

dyzhuu commented:
I think it makes a bit more sense for the route to be named something like `admin/semesters`

tulza committed:
css only empty schedules

tulza committed:
init popover refator

tulza committed:
style empty session text

tulza committed:
Merge branch 'uabc-178-Creation-of-edit-schedules-dashboard' of https://github.com/UoaWDCC/uabc-portal into uabc-178-Creation-of-edit-schedules-dashboard

tulza committed:
popover component

tulza committed:
fixed popover open classes

dyzhuu commented on src/components/popover.tsx:
```diff
@@ -0,0 +1,78 @@
+  );
+};
+
+export const PopoverOpen = ({ children }: { children: ReactNode }) => {
+  const { handleOpen } = useContext(PopoverContext);
+  return (
+    <div className="w-full h-full" onClick={handleOpen}>
+      {children}
+    </div>
+  );
+};
+
+export const PopoverClose = ({ children }: { children: ReactNode }) => {
+  const { handleClose } = useContext(PopoverContext);
+  return (
+    <div className="w-full h-full" onClick={handleClose}>
+      {children}
+    </div>
+  );
+};
```
are these being used?

dyzhuu commented on src/components/admin/SemesterDetailCard.tsx:
```diff
@@ -0,0 +1,79 @@
+  breakStart,
+  breakEnd,
+}: DetailCardProps) => {
+  const SemesterPopover = () => {
+    return (
+      <PopoverContainer>
+        <EditButton />
+        <DeleteButton />
+      </PopoverContainer>
+    );
+  };
```
Don't define a component inside another component. This compoenent is re-created everytime the out component is re-rendered which is very bad for performance reasons

dyzhuu commented on src/components/popover.tsx:
```diff
@@ -0,0 +1,78 @@
+  handleClose: () => void;
+};
+
+const PopoverContext = createContext<PopoverContextType>({});
```
typescript error

dyzhuu commented on src/components/admin/SemesterDetailCard.tsx:
```diff
@@ -0,0 +1,79 @@
+          className="absolute top-0 left-0 w-full h-full z-10"
+        />
+
+        {/* <input
+          type="radio"
+          name="session"
+          className="absolute outline top-0 left-0 z-10 w-full h-full bg- opacity-0 cursor-pointer"
+          value="id"
+        /> */}
```
remove comment

dyzhuu commented on src/components/admin/SemesterDetailCard.tsx:
```diff
@@ -0,0 +1,79 @@
+  breakStart: string;
+  breakEnd: string;
+};
+export const ScheduleContext = createContext<DetailCardProps>({});
```
typescript error

dyzhuu commented:
I think that given the scope of your components, useContext is actually good and won't cause many unnecessary re-renders.

However, if you're interested in how zustand could be used you can look at this
https://github.com/pmndrs/zustand/blob/main/docs/guides/initialize-state-with-props.md

dyzhuu commented on src/components/admin/EditDialogue.tsx:
Currently the EditDialogue & EditForm component is tied to the SemesterDetailsCard. Can you think about how you can make it composable & reusuable so that these components can be other for other components with similar UI needs?

dyzhuu commented on src/components/popover.tsx:
```diff
@@ -0,0 +1,78 @@
+const PopoverContext = createContext<PopoverContextType>({});
+
+export const Popover = ({ children }: { children: ReactNode }) => {
+  const [open, isOpen] = useState(false);
```
setOpen?

dyzhuu commented on src/components/popover.tsx:
maybe close the popover when `esc` is pressed & close it when either `Edit Sessions` or `Delete Sessions` is pressed?

dyzhuu commented on src/components/admin/EditDialogue.tsx:
```diff
@@ -0,0 +1,108 @@
+  const onSubmit: SubmitHandler<UserResponse> = (data) => console.log(data);
+
+  return (
+    <>
```
Do you need this fragment

dyzhuu commented on src/components/admin/EditDialogue.tsx:
```diff
@@ -0,0 +1,108 @@
+} from "../ui/dialog";
+import { ScheduleContext } from "./SemesterDetailCard";
+
+type UserResponse = {
```
User Response is really vague, can you rename this to something more specific? Like EditSemesterFormValues

dyzhuu commented on src/components/admin/EditDialogue.tsx:
```diff
@@ -0,0 +1,108 @@
+    <>
+      <form
+        className="flex gap-4 flex-col"
+        onSubmit={(e) => {
+          handleSubmit((data) => {
+            console.log("submitted");
+            onSubmit(data);
+          })(e);
+        }}
```
```suggestion
        onSubmit={handleSubmit(onSubmit)}
```

And do any additional stuff in the onSubmit function instead

dyzhuu commented on src/components/admin/EditDialogue.tsx:
```diff
@@ -0,0 +1,108 @@
+    handleSubmit,
+    formState: { errors },
+  } = useForm<UserResponse>({
+    defaultValues: {
+      startDate: startDate,
+      endDate: endDate,
+      breakStart: breakStart,
+      breakEnd: breakEnd,
+    },
```
```suggestion
    defaultValues: {
      startDate,
      endDate,
      breakStart,
      breakEnd,
    },
```

dyzhuu commented on src/components/admin/EditDialogue.tsx:
Could you also display the error text
![image](https://github.com/UoaWDCC/uabc-portal/assets/119282717/3459f1a9-bd58-4fe6-9eff-cdfebdad210c)

dyzhuu commented on src/app/admin/edit-sessions/page.tsx:
```diff
@@ -0,0 +1,55 @@
+  return response.json();
+};
+
+const page = async () => {
```
Maybe SemestersPage?

dyzhuu commented on src/components/TextInput.tsx:
```diff
@@ -12,34 +12,36 @@ interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
+    { label, type, isError, className, ...props }: InputProps,
+    ref: LegacyRef<HTMLInputElement>,
+  ) => {
+    return (
```
```suggestion
export const TextInput = forwardRef<HTMLInputElement, TextInputProps>(
  ({ label, type, isError, className, ...props }, ref) => {
```

tulza replied:
this would cause a typescript error at ref={ref} no?

dyzhuu replied:
What's the typescript error? Shows up fine for me

tulza committed:
fixed some conversation + testing api for semester

tulza committed:
Merge branch 'main' into uabc-178-Creation-of-edit-schedules-dashboard

tulza committed:
creation of semestersList

tulza committed:
Skeleton semester card

tulza committed:
dialog form date validation and error displaying

tulza committed:
memo'd semester card and deleted old unused component

tulza committed:
popover exit on esc

tulza committed:
fixed date validation and dialog styling

tulza committed:
refactoring dialog card, dialog utils, and many style changes

dyzhuu commented on src/lib/utils.ts:
```diff
@@ -33,3 +33,18 @@ export function convertTo12HourFormat(time24: string): string {
   // Return the formatted time string
   return `${hours}:${formattedMinutes}${period}`;
 }
+
+export const validateDate = (date: string) => {
+  // format DD/MM/YY
+  const dateParts: string[] = date.split("/");
+  //simple date validation
+  if (dateParts.length != 3 || dateParts.includes("")) return false;
+  if (2 < dateParts[0].length) return false; //DD
+  if (2 < dateParts[1].length) return false; //MM
+  if (dateParts[2].length != 4) return false; //YY
+
+  // Date("year/month/day"),
+  const dateStr = +dateParts[2] + "/" + +dateParts[1] + "/" + +dateParts[0];
+  const dateObject = new Date(dateStr);
+  return dateObject.toString() != "Invalid Date";
+};
```
consider using zod to validate this instead
https://zod.dev/?id=dates

https://github.com/react-hook-form/resolvers?tab=readme-ov-file#zod

dyzhuu commented on src/components/admin/SemesterEditDialog.tsx:
```diff
@@ -0,0 +1,101 @@
+        <div className="flex gap-2 *:grow ">
+          <TextInput
+            label="Start Date"
+            type="input"
```
type="text"

dyzhuu commented on src/components/admin/SemestersList.tsx:
```diff
@@ -0,0 +1,50 @@
+import SkeletonSemesterCard from "./SkeletonSemesterCard";
+
+export const SemestersList = () => {
+  const { data, isLoading } = useCurrentSemesters();
```
Fetch data in parent server component & pass down as initialData

dyzhuu commented on src/app/api/semesters/route.ts:
what's wrong with this?

dyzhuu commented on src/app/api/semesters/route.ts:
```diff
@@ -9,13 +9,13 @@ import { insertSemesterSchema } from "@/lib/validators";
-    if (user.role != "admin") {
-      return new Response("ERROR: No valid permissions", { status: 403 });
-    }
+    // const user = await getCurrentUser();
+    // if (!user) {
+    //   return new Response("ERROR: Unauthorized request", { status: 401 });
+    // }
+    // if (user.role != "admin") {
+    //   return new Response("ERROR: No valid permissions", { status: 403 });
+    // }
```
make sure to uncomment this later

dyzhuu commented on src/components/admin/SemesterDetailCard.tsx:
```diff
@@ -0,0 +1,69 @@
+  );
+};
+
+const SemesterPopover = () => {
+  return (
+    <Popover>
+      <PopoverOpenButton variant="outline" className="w-8 h-6 z-10">
+        <Ellipsis className="stroke-tertiary absolute w-4" />
+      </PopoverOpenButton>
+      <PopoverContainer>
+        <EditButton />
+        <DeleteButton />
+      </PopoverContainer>
+    </Popover>
+  );
+};
```
re: making this reusuable instead of creating new Popover / Edit / Delete buttons everytime we need it.

I think you can come back to this after you fully finish the edit and delete implementation

but an idea is passing the form as a prop

dyzhuu commented:
For revalidating data look at

https://tanstack.com/query/latest/docs/framework/react/guides/query-invalidation
https://tanstack.com/query/latest/docs/framework/react/guides/invalidations-from-mutations
https://tanstack.com/query/latest/docs/framework/react/guides/optimistic-updates

dyzhuu commented:
you could fetch the data in a server component:
https://nextjs.org/docs/app/building-your-application/data-fetching/patterns

integrating with tanstack query:
https://tanstack.com/query/latest/docs/framework/react/guides/ssr

tulza committed:
refactor options button to reusable

dyzhuu commented on src/components/admin/SemestersList.tsx:
```diff
@@ -0,0 +1,50 @@
+
+import { useCurrentSemesters } from "@/hooks/query/useCurrentSemesters";
+import { SemesterDetailCard } from "./SemesterDetailCard";
+import SkeletonSemesterCard from "./SkeletonSemesterCard";
```
make this a named export

tulza committed:
replace validation with zod and reset input on close dialog

tulza committed:
Co-authored-by: David Zhu <dyzhuu@users.noreply.github.com>

tulza committed:
datefns for parsing, invalidate queries and semester context refactor

tulza committed:
Merge branch 'main' into uabc-178-Creation-of-edit-schedules-dashboard

tulza committed:
toast ui

tulza committed:
toast ui

dyzhuu committed:
Fix enum issue

dyzhuu committed:
Merge branch 'uabc-178-Creation-of-edit-schedules-dashboard' of https://github.com/UoaWDCC/uabc-portal into uabc-178-Creation-of-edit-schedules-dashboard

tulza committed:
Add Toast popover on error and success

-Added root toaster to page.tsx
-Removed testing toaster in semesterlist.tsx

tulza committed:
Remove text bg on shadcn component

-text bg does not match the toaster so i had to dug into the component
to find and remove the single bg tw :)

tulza committed:
Refactor semester edit

- sort semester data by name
- fixed date format of body for PUT
- change text for toast
- added parse date to utils

tulza committed:
Add Tanstack mutation + multiple refactors

- modified dialog to include handleclose function
- use mutation for fetching
- success toast on onsuccess
- error toast on onerror
- update dialog data on close after confirm
- replaced old validate date and parse date

tulza committed:
Refactor OptionDialogPopoverBase for functionality

- Added a context for option dialog component to allow dialog closing by
  function
- Modified edit dialog to close on succuess

tulza committed:
Added dialog inputfield + improve code readability

- Added DialogInputField component for dialog inputfield
- Documentation to utils, dialogUtils, and admin page

dyzhuu commented on src/components/DialogUtils.tsx:
Eric did something similar to this in his ticket with adding a error/success message text field, do you want to talk to him about this and find a common solution?

dyzhuu commented on src/app/admin/semesters/page.tsx:
```diff
@@ -0,0 +1,17 @@
+      <div className="gap-4 flex flex-col empty:grow mb-4 empty:after:content-['No_semesters_set'] empty:after:text-tertiary empty:after:w-full empty:after:h-full empty:after:grid empty:after:place-items-center empty:after:grow empty:after:text-lg empty:after:font-medium">
+        <SemestersList />
+        <Toaster /> {/* Shadcn global toaster */}
+      </div>
```
You should put the toaster in a global context. Maybe under the main layout.tsx where the providers are? You can check out the other ticket i'm working on to see where I put it.

https://github.com/UoaWDCC/uabc-portal/blob/UABC-181-create-admin-view-session-card-component/src/app/layout.tsx

dyzhuu commented on src/app/admin/semesters/page.tsx:
```diff
@@ -0,0 +1,17 @@
+const SemestersPage = () => {
+  return (
+    <div className="relative flex px-4 min-h-dvh max-w-dvw flex-col bg-background dark overflow-x-hidden">
+      <EditSessionsHeader />
```
Oops, I think there's been a bit of desync and lack of communication between the different branches, we actually have a `<NavigationBar>` component for this component that was created in another PR.

tulza replied:
oops

dyzhuu replied:
So can we delete `EditSessionsHeader` and change this to use the `NavigationBar` component instead?

dyzhuu commented on src/components/admin/SemestersList.tsx:
```diff
@@ -0,0 +1,49 @@
+  ));
+};
+
+export default SemestersList;
```
No default exports for components

dyzhuu commented on src/components/admin/SemesterDetailCard.tsx:
```diff
@@ -0,0 +1,45 @@
+        <h3 className="text-lg text-foreground whitespace-nowrap">{name}</h3>
+        <OptionsPopoverButton>
+          <OptionDialogPopoverBase
+            ButtonElem={<OptionButtonUtils type="edit" />}
```
full names rather than abbreviations please
`ButtonElement`
`DialogElement`

dyzhuu commented on src/components/OptionButton.tsx:
```diff
@@ -0,0 +1,18 @@
+import React, { type ReactNode } from "react";
```
consider exporting related components under an `index.ts`

Take a look at how its done under `components/booking/SelectSessionList` etc.

This can be done in multiple places, such as with your SemesterComponents.

You can also consider using the compound component pattern in cases where you think it makes sense.
https://medium.com/@vitorbritto/react-design-patterns-compound-component-pattern-ec247f491294#:~:text=Compound%20components%20are%20a%20pattern,state%20and%20handle%20logic%20together.

I think that the Popover component can adopt this pattern:

```
<Popover>
 <Popover.Button>
    ...
 </Popover.Button>
 <Popover.Container>
    ...
 </Popover.Container>
</Popover>
```

dyzhuu committed:
chore: Refactor semester edit form using hookform zod resolver

dyzhuu commented on src/components/DialogUtils.tsx:
```diff
@@ -0,0 +1,73 @@
+  errorMessage?: string;
+}
+
+export const DialogInputField = forwardRef<
```
Any reason for creating a new component for this? Maybe we can incorporate the error message into the existing TextInput component? Talk to @monoclonalAb about this

tulza committed:
Refactor TextInput + Resolve conversations

- added the params errorMessage to textinput
- move toast to layout
- documentation in option popover

tulza committed:
Refactor popover & options to use compound pattern

- Combined OptionsPopover and OptionButton together
- Renamed OptionItemPopover to OptionsPopover
- Refactored Popover and Options to use Compound pattern
- Improved documentations over components
- slight modification to toast to have destruction variant to have only
  white text
- Merge DialogTextInput into TextInput
- add IsSuccess to textinput be compatible to @monoclonalAb pr

dyzhuu commented on src/components/DialogUtils.tsx:
```diff
@@ -47,27 +46,3 @@ export const DialogCardFooter = ({ ...props }: ButtonProps) => {
 interface DialogInputFieldProps extends InputProps {
   errorMessage?: string;
 }
```
this is no longer being used right?

dyzhuu commented on src/app/api/schedules/[scheduleId]/route.ts:
```diff
@@ -65,7 +66,18 @@ export async function PUT(
     }
 
-    if (updatedGameSession.startTime >= updatedGameSession.endTime) {
+    const parsedStartTime = parse(
+      updatedGameSession.startTime,
+      "dd/MM/yyyy",
+      new Date(),
+    );
+    const parsedEndTime = parse(
+      updatedGameSession.startTime,
+      "dd/MM/yyyy",
+      new Date(),
+    );
+    if (parsedStartTime >= parsedEndTime) {
+      console.log("parsedStartTime >= parsedEndTime");
```
remove this console log

dyzhuu replied:
I'm not sure if you're editing this right thing here. startTime and endTime are time strings i.e. "09:00:00", not date strings. Also this code pertains to gameSessionSchedules, but you're working with semesters.

This did bring to my attention that certain routes are missing correct validation. I will make a refactor ticket for that but you can delete this change.

dyzhuu commented on src/components/admin/SemesterEditDialog.tsx:
```diff
@@ -0,0 +1,159 @@
+      onSuccess: () => {
+        queryClient.invalidateQueries({ queryKey: ["semesters"] });
+        toast({
+          description: "Success!",
```
Can we make the toasts more descriptive? I just chatgpted this but for example you could have something like

```
toast({
    title: "Uh oh! Something went wrong",
    description:
    "An error occurred while updating the semester. Please try again.",
    variant: "destructive",
});
```

and 

```
toast({
    title: "Success!",
    description: "Semester details successfully updated",
});
```

dyzhuu commented on src/app/admin/semesters/page.tsx:
```diff
@@ -0,0 +1,15 @@
+
+const SemestersPage = () => {
+  return (
+    <div className="relative flex px-4 min-h-dvh max-w-dvw flex-col bg-background dark overflow-x-hidden">
```
Can you remove the dark mode styling?

dyzhuu commented on src/components/admin/SemesterEditDialog.tsx:
```diff
@@ -0,0 +1,159 @@
+
+    mutation.mutate(body, {
+      onError: (e) => {
+        console.log(e);
```
No console logs in production

dyzhuu commented:
Can we also validate that:

`startDate < breakStart`
`breakStart < breakEnd`
`breakEnd < endDate`

You can use zod's `.refine()` or `.superRefine()` for this.

https://zod.dev/?id=refine

dyzhuu commented on src/components/admin/SemesterEditDialog.tsx:
```diff
@@ -0,0 +1,159 @@
+import { useToast } from "../ui/use-toast";
+import { SemesterContext } from "./SemestersContext";
+
+const formSchema = z.object({
+  startDate: z
+    .string()
+    .min(1, "Field is required")
+    .refine(validateDate, "Invalid date"),
+  endDate: z
+    .string()
+    .min(1, "Field is required")
+    .refine(validateDate, "Invalid date"),
+  breakStart: z
+    .string()
+    .min(1, "Field is required")
+    .refine(validateDate, "Invalid date"),
+  breakEnd: z
+    .string()
+    .min(1, "Field is required")
+    .refine(validateDate, "Invalid date"),
+});
+
```
Can we also validate that:

`startDate < breakStart`
`breakStart < breakEnd`
`breakEnd < endDate`

You can use zod's `.refine()` or `.superRefine()` for this.

https://zod.dev/?id=refine

tulza committed:
Add Zod validation to form

- Created new utils function to handle comparing two dates
- Validate form using function above
- modify TextInput error message for better UX
- remove bad parse in api
- remove dark mode tw on page

tulza committed:
Refactor TextInput errorMessage to do layout shift

- Refactored error msg container to a block so it can do layout shift
- Animated the layout shift so it doesn't look weird
- Added tailwind transition property max height
- Minor fix: remove dark class on dialog

tulza committed:
Change error message anima duration reduce + merge

- animatoin duration 500 -> 150ms
- Merge branch 'main' into uabc-178-Creation-of-edit-schedules-dashboard

tulza committed:
Merge to main + reduce animation duration

- Merge branch 'main' into uabc-178-Creation-of-edit-schedules-dashboard
- animatoin duration 500 -> 150ms
i Changes to be committed:

tulza committed:
chores: organised files into different folders

tulza committed:
chores: something

dyzhuu committed:
fix: faulty redirect breaking e2e test

tulza reviewed (commented):
‎

tulza reviewed (commented):
w

tulza committed:
Modify context in optionsPopover and semesters

- created useSemesterContext and OptionsDialogContext to give meaningful
  error instead of using useContext

dyzhuu reviewed (dismissed):
lgtm

tulza committed:
semesterDetailCard now have truncated name
